plugins/gpio/fu-plugin-gpio.c

```diff
@@ -0,0 +1,163 @@
+	}
+	g_set_error(error,
+		    G_IO_ERROR,
+		    G_IO_ERROR_INVALID_DATA,
+		    "cannot parse level, got %s and expected high|low",
+		    str);
+	return FALSE;
+}
+
+static gboolean
+fu_plugin_gpio_process_quirk(FuPlugin *self, const gchar *str, GError **error)
+{
+	FuPluginData *data = fu_plugin_get_data(self);
+	gboolean value = FALSE;
+	FuDevice *device;
+	g_auto(GStrv) split = g_strsplit(str, ",", -1);
+	g_autoptr(FuDeviceLocker) locker = NULL;
+
+	/* already set by previous GUID */
+	if (data->gpio_device != NULL)
+		return TRUE;
+
+	if (g_strv_length(split) != 3) {
+		g_set_error(error,
+			    G_IO_ERROR,
+			    G_IO_ERROR_INVALID_DATA,
+			    "invalid format, CHIP_NAME,PIN_NAME,LEVEL, got '%s'",
+			    str);
+		return FALSE;
+	}
+	if (!fu_plugin_gpio_parse_level(split[2], &value, error))
+		return FALSE;
+	device = fu_plugin_cache_lookup(self, split[0]);
+	if (device == NULL) {
+		g_set_error(error,
+			    FWUPD_ERROR,
+			    FWUPD_ERROR_NOT_FOUND,
+			    "GPIO device %s not found",
+			    split[0]);
+		return FALSE;
+	}
+	locker = fu_device_locker_new(device, error);
+	if (locker == NULL)
+		return FALSE;
+	g_set_object(&data->gpio_device, FU_GPIO_DEVICE(device));
+	if (!fu_gpio_device_assign(data->gpio_device, split[1], value, error)) {
+		g_prefix_error(error, "failed to assign %s: ", split[0]);
+		return FALSE;
+	}
+	return TRUE;
+}
+
+static gboolean
+fu_plugin_gpio_prepare(FuPlugin *self, FuDevice *device, FwupdInstallFlags flags, GError **error)
+{
+	GPtrArray *guids = fu_device_get_guids(device);
+	for (guint i = 0; i < guids->len; i++) {
+		const gchar *guid = g_ptr_array_index(guids, i);
+		const gchar *str;
+		str = fu_context_lookup_quirk_by_id(fu_plugin_get_context(self),
+						    guid,
+						    "GpioForUpdate");
+		if (str == NULL)
+			continue;
+		if (!fu_plugin_gpio_process_quirk(self, str, error))
+			return FALSE;
+	}
+	return TRUE;
+}
+
+static gboolean
+fu_plugin_gpio_cleanup(FuPlugin *self, FuDevice *device, FwupdInstallFlags flags, GError **error)
+{
+	FuPluginData *data = fu_plugin_get_data(self);
+
+	/* nothing was set */
+	if (data->gpio_device == NULL)
+		return TRUE;
+
+	/* close the fd we opened during ->prepare */
+	if (!fu_gpio_device_unassign(data->gpio_device, error)) {
+		g_prefix_error(error, "failed to unassign %s: ", fu_device_get_logical_id(device));
+		return FALSE;
+	}
+	g_clear_object(&data->gpio_device);
+	return TRUE;
+}
+
+static gboolean
+fu_plugin_gpio_device_created(FuPlugin *self, FuDevice *device, GError **error)
+{
+	g_autoptr(FuDeviceLocker) locker = NULL;
+
+	/* run setup */
+	locker = fu_device_locker_new(device, error);
+	if (locker == NULL)
+		return FALSE;
+
+	/* add to cache, which maintains a ref */
+	fu_plugin_cache_add(self, fu_device_get_logical_id(device), device);
+
+	/* we do *not* want to export this device over D-Bus */
+	g_set_error_literal(error,
+			    FWUPD_ERROR,
+			    FWUPD_ERROR_NOT_SUPPORTED,
+			    "not exporting GPIO device");
+	return FALSE;
```

Comment:
this feels pretty hacky.  Why can't it export and just not be an updatable device?  Like the CPU is.

Reply:
Hmm, I had it that way before. Maybe it makes it better to export.